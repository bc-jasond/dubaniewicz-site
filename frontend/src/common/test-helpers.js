import { fromJS } from 'immutable';
import { reviver } from './utils';
import { SELECTION_LENGTH, SELECTION_NEXT } from './constants';

export function overrideConsole() {
  global.console = {
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn()
  };
}

export function mockLocalStorage() {
  global.localStorageStorage = {};
  global.localStorage = {
    getItem: k => global.localStorageStorage[k],
    setItem: (k, v) => (global.localStorageStorage[k] = v),
    clear: () => (global.localStorageStorage = {})
  };
}

export function mockSetTimeout() {
  global.setTimeout = jest.fn();
}

export function makeSelections(values) {
  const head = {};
  let current = head;
  let prev;
  do {
    let [currentLength, ...currentValues] = values.shift();
    current[SELECTION_LENGTH] = currentLength;
    currentValues.forEach(v => {
      if (typeof v === 'object') {
        current[v.key] = v.value;
      } else {
        current[v] = true;
      }
    });
    if (prev) {
      prev[SELECTION_NEXT] = current;
    }
    prev = current;
    current = {};
  } while (values.length);
  prev[SELECTION_NEXT] = undefined; //important for the reviver()
  prev[SELECTION_LENGTH] = -1;
  return fromJS(head, reviver);
}
