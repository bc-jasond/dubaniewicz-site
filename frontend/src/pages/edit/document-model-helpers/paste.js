import { handlePasteTextType } from './handle-text-type';

export function doPaste(documentModel, selectionOffsets, clipboardData) {
  const [[caretPositionStart, _, selectedNodeId]] = selectionOffsets;

  if (!selectedNodeId) {
    return [];
  }

  // split selectedNodeContent at caret
  const clipboardText = clipboardData.getData('text/plain');

  if (!documentModel.isTextType(selectedNodeId)) {
    console.warn(
      'doPaste() - trying to paste into a MetaType node is not supported',
      selectionOffsets,
      clipboardData
    );
    return [];
  }
  return handlePasteTextType(
    documentModel,
    selectedNodeId,
    caretPositionStart,
    clipboardText
  );
}
