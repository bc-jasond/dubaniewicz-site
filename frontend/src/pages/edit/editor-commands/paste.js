/* eslint-disable import/prefer-default-export */
import { assertValidDomSelectionOrThrow } from '../../../common/dom';
import {
  adjustSelectionOffsetsAndCleanup,
  splitSelectionsAtCaretOffset,
} from '../selection-helpers';
import { getLastExecuteIdFromHistory } from '../history-manager';

export function doPaste(documentModel, selectionOffsets, clipboardData) {
  assertValidDomSelectionOrThrow(selectionOffsets);
  const { caretStart, startNodeId } = selectionOffsets;

  // split selectedNodeContent at caret
  const clipboardText = clipboardData.getData('text/plain');
  // TODO: convert HTML => filbert documentModel
  // const clipboardHtml = clipboardData.getData('text/html');

  if (!documentModel.isTextType(startNodeId)) {
    console.warn(
      `doPaste() - trying to paste into a "${documentModel
        .getNode(startNodeId)
        .get('type')}" node is not supported`,
      selectionOffsets,
      clipboardData
    );
    return {};
  }

  const historyState = [];
  const getReturnPayload = (executeSelectionOffsets) => {
    return { executeSelectionOffsets, historyState };
  };

  // TODO: preserve format of copied text
  let selectedNode = documentModel.getNode(startNodeId);
  const content = selectedNode.get('content') || '';
  const contentLeft = content.substring(0, caretStart);
  const contentRight = content.substring(caretStart);
  console.info(
    'PASTE - paragraph content: ',
    contentLeft,
    contentRight,
    caretStart,
    clipboardText
  );
  const clipboardLines = clipboardText.split('\n');
  // single line paste
  if (clipboardLines.length === 1) {
    const updatedContent = `${contentLeft}${clipboardText}${contentRight}`;
    const { length: diffLength } = clipboardText;
    selectedNode = selectedNode.set('content', updatedContent);
    selectedNode = adjustSelectionOffsetsAndCleanup(
      selectedNode,
      content,
      caretStart,
      diffLength
    );
    historyState.push(...documentModel.update(selectedNode));
    return getReturnPayload({
      startNodeId,
      caretStart: contentLeft.length + diffLength,
    });
  }
  // MULTI LINE PASTE
  // NOTE: the order of operations is important here
  console.info('PASTE: multi-line');
  const firstLine = clipboardLines.shift();
  const lastLine = clipboardLines.pop();
  const leftNodeId = startNodeId;
  // add 'lastLine' content to beginning of "right"
  const updatedRightNodeContent = `${lastLine}${contentRight}`;
  // add 'firstLine' content to end of "left"
  const updatedLeftNodeContent = `${contentLeft}${firstLine}`;
  // TODO: this is largely copied from handleEnterTextType() above, maybe split the shared code into a helper?
  // NOTE: rightNode insert (last before first) so that leftNode gets updated with a next_sibling_id
  // 1) insert a new node (rightNode) for 'lastLine' - using BEFORE content
  historyState.push(
    ...documentModel.insert(selectedNode.get('type'), leftNodeId, contentRight)
  );
  const rightNodeId = getLastExecuteIdFromHistory(historyState);
  // now leftNode has 'next_sibling_id' set to rightNode
  // important: 'content' is now contentLeft
  let leftNode = documentModel.getNode(leftNodeId).set('content', contentLeft);
  let rightNode = documentModel.getNode(rightNodeId);

  // 2) if the original selected node can have Selections - move them to the right node if needed
  // NOTE: do this with BEFORE content
  ({ leftNode, rightNode } = splitSelectionsAtCaretOffset(
    leftNode,
    rightNode,
    caretStart
  ));

  // 3) update content with firstLine & lastLine
  leftNode = leftNode.set('content', updatedLeftNodeContent);
  rightNode = rightNode.set('content', updatedRightNodeContent);

  // 4) adjust offsets with updated content
  leftNode = adjustSelectionOffsetsAndCleanup(
    leftNode,
    contentLeft, // this is before content! takes this as an argument for comparison with now updated content in leftNode
    caretStart,
    firstLine.length
  );
  rightNode = adjustSelectionOffsetsAndCleanup(
    rightNode,
    contentRight, // before content!
    0,
    lastLine.length
  );
  historyState.push(...documentModel.update(leftNode));

  // there are middle lines, insert Paragraphs after "left"
  let prevNodeId = startNodeId;
  while (clipboardLines.length > 0) {
    const currentLine = clipboardLines.shift();
    // skip whitespace only lines TODO: allow in Code sections?
    if (currentLine.trim().length > 0) {
      historyState.push(
        ...documentModel.insert(leftNode.get('type'), prevNodeId, currentLine)
      );
      const nextId = getLastExecuteIdFromHistory(historyState);
      prevNodeId = nextId;
    }
  }

  historyState.push(...documentModel.update(rightNode));
  return getReturnPayload({
    startNodeId: rightNodeId,
    caretStart: lastLine.length,
  });
}
