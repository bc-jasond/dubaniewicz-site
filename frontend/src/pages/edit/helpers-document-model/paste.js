import { NODE_TYPE_LI, NODE_TYPE_P, NODE_TYPE_PRE } from '../../../common/constants';
import { getNodeId, getNodeType } from '../../../common/dom';
import { handlePasteCode } from '../helpers-by-section-type/handle-code';
import { handlePasteParagraph } from '../helpers-by-section-type/handle-paragraph';

export function doPaste(documentModel, selectionOffsets, clipboardData) {
  const [
    [caretPositionStart, _, selectedNode],
  ] = selectionOffsets;
  
  if (!selectedNode) {
    return [];
  }
  const selectedNodeType = getNodeType(selectedNode);
  const selectedNodeId = getNodeId(selectedNode);
  
  // split selectedNodeContent at caret
  const clipboardText = clipboardData.getData('text/plain');
  
  let focusNodeId;
  let focusIdx = caretPositionStart;
  switch (selectedNodeType) {
    case NODE_TYPE_PRE: {
      [
        focusNodeId,
        focusIdx,
      ] = handlePasteCode(documentModel, selectedNodeId, caretPositionStart, clipboardText);
      break;
    }
    case NODE_TYPE_LI: // fall-through
    case NODE_TYPE_P: {
      [
        focusNodeId,
        focusIdx,
      ] = handlePasteParagraph(documentModel, selectedNodeId, caretPositionStart, clipboardText);
    }
    default: { /*NOOP*/
    }
  }
  return [focusNodeId, focusIdx];
}
