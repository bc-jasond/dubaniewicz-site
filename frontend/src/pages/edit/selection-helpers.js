import { fromJS, Map } from 'immutable';

import { SELECTION_LENGTH, SELECTION_NEXT } from '../../common/constants';
import {
  cleanTextOrZeroLengthPlaceholder,
  reviver,
  Selection
} from '../../common/utils';

export function formatSelections(head) {
  if (!Map.isMap(head)) {
    return '';
  }
  let current = head;
  let output = 'head: ';
  while (current) {
    output = `${output}${current.get(SELECTION_LENGTH)}, ${current.get(
      SELECTION_NEXT
    )} | `;
    current = current.get(SELECTION_NEXT);
  }
  return output;
}

/**
 * PRIVATE HELPERS
 */
// returns "head" or first selection
// convert to JS to avoid having to refresh all references to "next" on every change
function getSelections(nodeModel) {
  return nodeModel.getIn(['meta', 'selections'], Selection()).toJS();
}
function setSelections(nodeModel, value) {
  return nodeModel.setIn(
    ['meta', 'selections'],
    value ? fromJS(value, reviver) : Selection()
  );
}

function selectionsHaveIdenticalFormats(left, right) {
  // remove non-formatting related fields
  const leftCompare = Selection(left)
    .delete(SELECTION_LENGTH)
    .delete(SELECTION_NEXT);
  const rightCompare = Selection(right)
    .delete(SELECTION_LENGTH)
    .delete(SELECTION_NEXT);
  // use built-in equals()
  return leftCompare.equals(rightCompare);
}
/* eslint-disable no-param-reassign, prefer-destructuring */
function maybeMergeNodes(left, right) {
  if (!selectionsHaveIdenticalFormats(left, right)) {
    return false;
  }
  // if we're past the end or on the last node
  if (!right || !right[SELECTION_NEXT]) {
    left[SELECTION_LENGTH] = -1;
    left[SELECTION_NEXT] = undefined;
  } else {
    left[SELECTION_LENGTH] += right[SELECTION_LENGTH];
    left[SELECTION_NEXT] = right[SELECTION_NEXT];
  }
  return true;
}
/* eslint-enable no-param-reassign, prefer-destructuring */

// uses vanilla JS instead of Immutable
function internalGetSelectionAtIndex(head, idx) {
  let selection = head;
  let i = 0;
  while (selection && i < idx) {
    // eslint-disable-next-line prefer-destructuring
    selection = selection[SELECTION_NEXT];
    i += 1;
  }
  if (!selection) {
    console.error('Bad selection index: ', idx, head);
  }
  return selection;
}

/**
 * PUBLIC API
 *
 *
 * Every public export calls this to keep shit on the level.
 * FOR POSTERITY: This has been the single most difficult function to design in the whole codebase
 * bugs and regressions here over and over again finally prompted adding the first tests with jest
 *
 * adjusts selection offsets (and removes selections) after these events: paste, keydown, delete 1 char, delete selection of 1 or more chars
 */
export function adjustSelectionOffsetsAndCleanup(
  nodeModel,
  beforeContent = '',
  start = 0,
  count = 0
) {
  const doesRemoveCharacters = count < 0;
  // no-op?
  if (start === 0 && count === 0) {
    return nodeModel;
  }
  // validate input
  if (
    // can't start before 0
    start < 0 ||
    // trying to delete too far left (past 0)
    start + count < 0 ||
    // trying to delete from too far right
    (doesRemoveCharacters && start > beforeContent.length) ||
    // trying to add too many characters
    (!doesRemoveCharacters && start + count > nodeModel.get('content')?.length)
  ) {
    throw new Error(
      `adjustSelectionOffsetsAndCleanup out of bounds!\n${JSON.stringify(
        nodeModel.toJS()
      )}\n${start}\n${count}`
    );
  }

  let head = getSelections(nodeModel);
  console.info(
    'ADJUST         ',
    formatSelections(head),
    ' -- offset: ',
    start,
    ' count: ',
    count
  );
  // if there's only 1 selection, no-op
  if (!head[SELECTION_NEXT]) {
    return nodeModel;
  }
  let current = head;
  let prev;
  let caretPosition = 0;
  // if we're ADDING content
  if (!doesRemoveCharacters) {
    // find the selection
    while (
      current[SELECTION_NEXT] &&
      start >= caretPosition + current[SELECTION_LENGTH]
    ) {
      caretPosition += current[SELECTION_LENGTH];
      // eslint-disable-next-line prefer-destructuring
      current = current[SELECTION_NEXT];
    }
    // and increase it's length if it's not at the end
    if (current[SELECTION_NEXT]) {
      current[SELECTION_LENGTH] += count;
      return setSelections(nodeModel, head);
    }
    return nodeModel;
  }
  //
  // if we're DELETING content:
  //
  // did we delete all content for the node? - count is negative
  if (beforeContent.length + count === 0) {
    return setSelections(nodeModel);
  }
  const deleteCaretStart = start + count; // count is negative
  if (deleteCaretStart === 0) {
    // unset head if we deleted through the beginning
    head = undefined;
  }
  // loop through all selections
  let didSkip = false;
  let didDelete = false;
  while (current) {
    // capture this value before we mutate it - give last selection a length for comparison (it will be -1)
    const currentLength =
      current[SELECTION_LENGTH] === -1
        ? beforeContent.length - caretPosition
        : current[SELECTION_LENGTH];
    if (
      caretPosition >= deleteCaretStart &&
      caretPosition + currentLength <= start
    ) {
      // whole selection was deleted
      didSkip = true;
    } else {
      // selection completely overlaps?
      if (
        caretPosition <= deleteCaretStart &&
        caretPosition + currentLength >= start
      ) {
        const newLength = currentLength + count;
        if (newLength === 0) {
          // deleted the selection
          didDelete = true;
        }
        current[SELECTION_LENGTH] = currentLength + count; // count is negative
      }
      // selection overlaps to the left?
      else if (
        caretPosition < deleteCaretStart &&
        caretPosition + currentLength > deleteCaretStart
      ) {
        current[SELECTION_LENGTH] = deleteCaretStart - caretPosition;
      }
      // selection overlaps to the right?
      else if (
        current[SELECTION_NEXT] &&
        caretPosition < start &&
        caretPosition + currentLength > start
      ) {
        current[SELECTION_LENGTH] = caretPosition + currentLength - start;
      }
      if (!head) {
        didSkip = false;
        head = current;
      } else if (didSkip) {
        didSkip = false;
        if (!selectionsHaveIdenticalFormats(prev, current)) {
          prev[SELECTION_NEXT] = current;
        } else if (current[SELECTION_NEXT]) {
          // merge new neighbors with same formats here
          prev[SELECTION_LENGTH] += current[SELECTION_LENGTH];
          // eslint-disable-next-line prefer-destructuring
          prev[SELECTION_NEXT] = current[SELECTION_NEXT];
        } else {
          // the last selection has same formats as prev and needs to be "merged" - just delete
          didDelete = true;
        }
      }
      if (didDelete) {
        // mark as skipped to link prev & current[next] on next loop
        didSkip = true;
        didDelete = false;
      } else {
        prev = current;
      }
    }
    // advance cursor and selection pointer
    caretPosition += currentLength;
    // eslint-disable-next-line prefer-destructuring
    current = current[SELECTION_NEXT];
  }
  // prev should be last node
  prev[SELECTION_LENGTH] = -1;
  prev[SELECTION_NEXT] = undefined;

  return setSelections(nodeModel, head);
}

/**
 * Takes a highlight range in paragraph content and maps it to a Selection.
 * finds index of existing selection, or replaces into the selections linked list a new placeholder selection
 * adjusting/removing existing overlapping selections in place
 * NOTE: doesn't merge neighboring selections with identical formats, this happens on update
 */
export function getSelectionByContentOffset(nodeModel, start, end) {
  // TODO: validation of start & end against nodeModel.get('content')?.length
  const length = end - start;
  const doesReplaceLastSelection = end === nodeModel.get('content', '').length;
  // first see if the exact Selection already exists?
  let head = getSelections(nodeModel);
  let originalHead = head;
  let current = head;
  let prev;
  let caretPosition = 0;
  let idx = 0;
  while (
    current[SELECTION_NEXT] &&
    caretPosition + current[SELECTION_LENGTH] <= start
  ) {
    caretPosition += current[SELECTION_LENGTH];
    idx += 1;
    prev = current;
    // eslint-disable-next-line prefer-destructuring
    current = current[SELECTION_NEXT];
  }
  // found exact match in existing selections ?
  if (
    caretPosition === start &&
    (current[SELECTION_LENGTH] === length ||
      (doesReplaceLastSelection && !current[SELECTION_NEXT]))
  ) {
    return { selections: fromJS(head, reviver), idx };
  }
  // if we're here, we didn't find an exact match in existing selections
  const newSelection = {
    [SELECTION_LENGTH]: doesReplaceLastSelection ? -1 : length,
    [SELECTION_NEXT]: undefined
  };
  // capture current length before mutation
  // eslint-disable-next-line prefer-destructuring
  const originalLength = current[SELECTION_LENGTH];
  // eslint-disable-next-line prefer-destructuring
  const originalNext = current[SELECTION_NEXT];
  // replace head?
  if (start === 0) {
    head = newSelection;
  } else if (start - caretPosition === 0) {
    // users selection lines up with end of last selection
    prev[SELECTION_NEXT] = newSelection;
  } else if (
    (!originalNext && !doesReplaceLastSelection) ||
    originalLength - (start - caretPosition) > length
  ) {
    // newSelection is completely inside of current
    const currentCopy = { ...current };
    current[SELECTION_LENGTH] = start - caretPosition;
    current[SELECTION_NEXT] = newSelection;
    idx += 1;
    newSelection[SELECTION_NEXT] = currentCopy;
    currentCopy[SELECTION_LENGTH] = !originalNext
      ? -1
      : originalLength - current[SELECTION_LENGTH] - length;
    return { selections: fromJS(head, reviver), idx };
  } else {
    // adjust "left" length
    current[SELECTION_LENGTH] = start - caretPosition;
    // no merge (happens on update), set left[next] to new selection
    current[SELECTION_NEXT] = newSelection;
    // move index pointer only if head didn't change
    idx += 1;
  }
  if (doesReplaceLastSelection) {
    return { selections: fromJS(head, reviver), idx };
  }
  // add original "left" length to caretPosition to preserve original selection lengths for further comparison
  caretPosition += originalLength;
  // advance pointer to continue through original list
  current = originalNext;
  // skip any completely overlapped selections
  while (current && caretPosition + current[SELECTION_LENGTH] <= end) {
    caretPosition += current[SELECTION_LENGTH];
    prev = current;
    // eslint-disable-next-line prefer-destructuring
    current = current[SELECTION_NEXT];
  }
  if (!prev) {
    // newSelection is new head
    // adjust originalHead length if newSelection length <
    if (
      originalHead[SELECTION_LENGTH] > 0 &&
      newSelection[SELECTION_LENGTH] < originalHead[SELECTION_LENGTH]
    ) {
      originalHead[SELECTION_LENGTH] -= newSelection[SELECTION_LENGTH];
    }
    head[SELECTION_NEXT] = originalHead;
    return { selections: fromJS(head, reviver), idx };
  }
  if (!current) {
    // newSelection is new 2nd to last, prev is last Selection
    newSelection[SELECTION_NEXT] = prev;
    return { selections: fromJS(head, reviver), idx };
  }
  if (current[SELECTION_NEXT]) {
    // adjust "right" length (if not -1)
    current[SELECTION_LENGTH] = caretPosition + current[SELECTION_LENGTH] - end;
  }
  // attach new selection to "right"
  newSelection[SELECTION_NEXT] = current;

  // TODO: apply overlapping formats?
  return { selections: fromJS(head, reviver), idx };
}

export function getSelectionAtIdx(head, idx) {
  let selection = head;
  let i = 0;
  while (selection && i < idx) {
    selection = selection.get(SELECTION_NEXT);
    i += 1;
  }
  if (!selection) {
    console.error('Bad selection index: ', idx, head);
  }
  return selection;
}

/**
 * insert a new formatted selection
 * NOTE: getSelectionByContentOffset() does the hard work of carving out a new selection
 *  this function just puts it back BUT, also merges neighboring selections
 *  with identical formats
 */
export function replaceSelection(nodeModelArg, newSelection, idx) {
  const nodeModel = nodeModelArg;
  let head = getSelections(nodeModel);
  const current = internalGetSelectionAtIndex(head, idx);
  const updated = newSelection.toJS();
  // eslint-disable-next-line prefer-destructuring
  updated[SELECTION_NEXT] = current[SELECTION_NEXT];
  if (idx === 0) {
    head = updated;
    maybeMergeNodes(head, head[SELECTION_NEXT]);
  } else {
    const prev = internalGetSelectionAtIndex(head, idx - 1);
    prev[SELECTION_NEXT] = updated;
    maybeMergeNodes(updated, updated[SELECTION_NEXT]);
    maybeMergeNodes(prev, updated);
  }

  return setSelections(nodeModel, head);
}

export function splitSelectionsAtCaretOffset(
  leftNodeModelArg,
  rightNodeModelArg,
  caretStart
) {
  const leftNode = leftNodeModelArg;
  const rightNode = rightNodeModelArg;
  const head = getSelections(leftNode);
  let current = head;
  let caretPosition = 0;
  while (
    current[SELECTION_NEXT] &&
    caretStart > caretPosition + current[SELECTION_LENGTH]
  ) {
    caretPosition += current[SELECTION_LENGTH];
    // eslint-disable-next-line prefer-destructuring
    current = current[SELECTION_NEXT];
  }
  const headRightLength = current[SELECTION_NEXT]
    ? current[SELECTION_LENGTH] - (caretStart - caretPosition)
    : -1;
  let headRight;
  if (headRightLength === -1) {
    // split in the middle of last selection
    headRight = current;
  } else if (headRightLength === 0) {
    // split on the edge of 2 selections
    // eslint-disable-next-line prefer-destructuring
    headRight = current[SELECTION_NEXT];
  } else {
    // splits in middle of a selection
    headRight = {
      ...current,
      [SELECTION_LENGTH]: headRightLength
    };
  }
  current[SELECTION_LENGTH] = -1;
  current[SELECTION_NEXT] = undefined;
  return {
    leftNode: setSelections(leftNode, head),
    rightNode: setSelections(rightNode, headRight)
  };
}

export function concatSelections(leftModelArg, rightModelArg) {
  const leftModel = leftModelArg;
  // the left last selection length will be -1, figure out it's new length
  let { length: leftLastSelectionLength } = leftModel.get('content');
  const rightModel = rightModelArg;
  const head = getSelections(leftModel);
  const headRight = getSelections(rightModel);
  let current = head;
  while (current[SELECTION_NEXT]) {
    leftLastSelectionLength -= current[SELECTION_LENGTH];
    // eslint-disable-next-line prefer-destructuring
    current = current[SELECTION_NEXT];
  }
  // current is now the last left selection
  current[SELECTION_LENGTH] = leftLastSelectionLength;
  // attach right selections
  current[SELECTION_NEXT] = headRight;
  // same formats? merge
  if (selectionsHaveIdenticalFormats(current, headRight)) {
    if (!headRight[SELECTION_NEXT]) {
      current[SELECTION_LENGTH] = -1;
      current[SELECTION_NEXT] = undefined;
    } else {
      current[SELECTION_LENGTH] += headRight[SELECTION_LENGTH];
      // eslint-disable-next-line prefer-destructuring
      current[SELECTION_NEXT] = headRight[SELECTION_NEXT];
    }
  }
  return setSelections(leftModel, head);
}

/**
 * get an array of content sliced-up by selection lengths
 */
export function getContentBySelections(node) {
  let content = node.get('content');
  if (content === undefined || content === null) {
    content = '';
  }
  const head = getSelections(node);
  // if there's no selection or just one, return the whole string
  if (!head[SELECTION_NEXT]) {
    return [cleanTextOrZeroLengthPlaceholder(content)];
  }
  const pieces = [];
  let caretPosition = 0;
  let current = head;
  while (current) {
    const end =
      current[SELECTION_LENGTH] > -1
        ? caretPosition + current[SELECTION_LENGTH]
        : undefined;
    // NOTE: content.substring(undefined, undefined) works like: content.substring(0, content.length)
    const piece = content.substring(caretPosition, end);
    pieces.push(cleanTextOrZeroLengthPlaceholder(piece));
    caretPosition += current[SELECTION_LENGTH];
    // eslint-disable-next-line prefer-destructuring
    current = current[SELECTION_NEXT];
  }
  return pieces;
}

export function getSelectionsLength(node) {
  const head = node.getIn(['meta', 'selections']);
  let i = 0;
  let current = head;
  while (current) {
    i += 1;
    current = current.get(SELECTION_NEXT);
  }
  return i;
}
