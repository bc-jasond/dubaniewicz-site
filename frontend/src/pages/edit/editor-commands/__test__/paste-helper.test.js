import { overrideConsole } from '../../../../common/test-helpers';
import DocumentModel from '../../document-model';
import * as selectionHelpers from '../../selection-helpers';
import { doPaste } from '../paste';
import {
  testPostWithAllTypesJS,
  imgId,
  firstNodeIdH1,
  formattedPContent,
  formattedPId,
} from '../../../../common/test-post-with-all-types';

const { post, contentNodes } = testPostWithAllTypesJS;
overrideConsole();
let doc = DocumentModel();

const mockText = 'some sample string';
const clipboardDataMock = {
  getData: () => mockText,
};

const spyAdjust = jest
  .spyOn(selectionHelpers, 'adjustSelectionOffsetsAndCleanup')
  .mockImplementation((...args) => args[0]);

beforeEach(() => {
  doc = DocumentModel(post.id, {}, contentNodes);
});
describe.skip('Document Model -> paste node helper', () => {
  test('doPaste - validates input', () => {
    const spy = jest
      .spyOn(textTypeHelpers, 'handlePasteTextType')
      .mockImplementation(() => {});
    // bad node id
    let noopResult = doPaste(doc, {}, clipboardDataMock);
    expect(noopResult).toEqual({});
    // meta type nodes not supported
    noopResult = doPaste(
      doc,
      { caretStart: 0, startNodeId: imgId },
      clipboardDataMock
    );
    expect(noopResult).toEqual({});
    expect(spy).not.toHaveBeenCalled();
  });
  test('doPaste - calls handlePasteTextType with expected parameters', () => {
    const caretStart = 10;
    const spy = jest
      .spyOn(textTypeHelpers, 'handlePasteTextType')
      .mockImplementation(() => {});
    doPaste(doc, { caretStart, startNodeId: firstNodeIdH1 }, clipboardDataMock);
    expect(spy).toHaveBeenCalledWith(doc, firstNodeIdH1, caretStart, mockText);
  });
  test('handlePasteTextType - plain text - single line paste', () => {
    const clipboardText = 'bib jibbs, bab jabs ';
    const caretStartArg = Math.floor(formattedPContent.length / 2);

    const { startNodeId, caretStart } = handlePasteTextType(
      doc,
      formattedPId,
      caretStartArg,
      clipboardText
    );
    expect(startNodeId).toBe(formattedPId);
    expect(caretStart).toBe(caretStartArg + clipboardText.length);
    expect(spyAdjust).toHaveBeenCalledWith(
      doc.getNode(formattedPId),
      formattedPContent,
      caretStartArg,
      clipboardText.length
    );
  });
  test('handlePasteTextType - plain text - multi line paste', () => {
    const clipboardText = 'bib jibbs\nbab jabs\nbubba rubba jubbson';
    const textLines = clipboardText.split('\n');
    const lastLine = textLines[textLines.length - 1];
    const caretStartArg = Math.floor(firstNodeIdH1.length / 2);
    const originalType = doc.getNode(firstNodeIdH1).get('type');

    const { startNodeId, caretStart } = handlePasteTextType(
      doc,
      firstNodeIdH1,
      caretStartArg,
      clipboardText
    );
    expect(startNodeId).not.toBe(firstNodeIdH1);
    expect(caretStart).toBe(lastLine.length);
    expect(spyAdjust).toHaveBeenCalledTimes(2);
    let current = doc.getNode(firstNodeIdH1);
    for (let i = 0; i < textLines.length; i++) {
      expect(current.get('type')).toBe(originalType);
      current = doc.getNextNode(current.get('id'));
    }
  });
});
