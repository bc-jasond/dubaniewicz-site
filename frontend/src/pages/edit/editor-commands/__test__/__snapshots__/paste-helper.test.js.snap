// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document Model -> paste node helper doPaste() pastes a single line of text 1`] = `
Array [
  Object {
    "executeState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "621e",
      "type": "h1",
      "content": "Large Headsome sample stringing",
      "meta": Immutable.Map {},
    },
    "unexecuteState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "621e",
      "type": "h1",
      "content": "Large Heading",
      "meta": Immutable.Map {},
    },
  },
]
`;

exports[`Document Model -> paste node helper doPaste() pastes plain text - multi line paste 1`] = `
Array [
  Object {
    "executeState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "3021",
      "type": "h1",
      "content": "Large Heading",
      "meta": Immutable.Map {},
    },
    "unexecuteState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "621e",
      "type": "h1",
      "content": "Large Heading",
      "meta": Immutable.Map {},
    },
  },
  Object {
    "executeState": Immutable.Map {
      "post_id": 175,
      "type": "h1",
      "content": "rge Heading",
      "meta": Immutable.Map {},
      "id": "3021",
      "next_sibling_id": "621e",
    },
    "unexecuteState": undefined,
  },
  Object {
    "executeState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "3021",
      "type": "h1",
      "content": "Labib jibbs",
      "meta": Immutable.Map {
        "selections": Immutable.Record {
          "next": undefined,
          "length": -1,
          "selection-bold": false,
          "selection-italic": false,
          "selection-code": false,
          "selection-siteinfo": false,
          "selection-mini": false,
          "selection-strikethrough": false,
          "selection-link": false,
          "linkUrl": "",
        },
      },
    },
    "unexecuteState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "3021",
      "type": "h1",
      "content": "Large Heading",
      "meta": Immutable.Map {},
    },
  },
  Object {
    "executeState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "1762",
      "type": "h1",
      "content": "Labib jibbs",
      "meta": Immutable.Map {
        "selections": Immutable.Record {
          "next": undefined,
          "length": -1,
          "selection-bold": false,
          "selection-italic": false,
          "selection-code": false,
          "selection-siteinfo": false,
          "selection-mini": false,
          "selection-strikethrough": false,
          "selection-link": false,
          "linkUrl": "",
        },
      },
    },
    "unexecuteState": Immutable.Map {
      "post_id": 175,
      "id": "8e34",
      "next_sibling_id": "3021",
      "type": "h1",
      "content": "Labib jibbs",
      "meta": Immutable.Map {
        "selections": Immutable.Record {
          "next": undefined,
          "length": -1,
          "selection-bold": false,
          "selection-italic": false,
          "selection-code": false,
          "selection-siteinfo": false,
          "selection-mini": false,
          "selection-strikethrough": false,
          "selection-link": false,
          "linkUrl": "",
        },
      },
    },
  },
  Object {
    "executeState": Immutable.Map {
      "post_id": 175,
      "type": "h1",
      "content": "bab jabs",
      "meta": Immutable.Map {},
      "id": "1762",
      "next_sibling_id": "3021",
    },
    "unexecuteState": undefined,
  },
  Object {
    "executeState": Immutable.Map {
      "post_id": 175,
      "type": "h1",
      "content": "bubba rubba jubbsonrge Heading",
      "meta": Immutable.Map {
        "selections": Immutable.Record {
          "next": undefined,
          "length": -1,
          "selection-bold": false,
          "selection-italic": false,
          "selection-code": false,
          "selection-siteinfo": false,
          "selection-mini": false,
          "selection-strikethrough": false,
          "selection-link": false,
          "linkUrl": "",
        },
      },
      "id": "3021",
      "next_sibling_id": "621e",
    },
    "unexecuteState": Immutable.Map {
      "post_id": 175,
      "type": "h1",
      "content": "rge Heading",
      "meta": Immutable.Map {},
      "id": "3021",
      "next_sibling_id": "621e",
    },
  },
]
`;
